File: platform/common-lib/src/main/java/com/researchex/common/security/InternalSecretFilter.java
package com.researchex.common.security;

import jakarta.servlet.FilterChain;
import jakarta.servlet.ServletException;
import jakarta.servlet.http.HttpServletRequest;
import jakarta.servlet.http.HttpServletResponse;
import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.util.Objects;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.util.AntPathMatcher;
import org.springframework.web.filter.OncePerRequestFilter;

/**
 * 내부 호출용 시크릿 검증을 수행하는 필터.
 * 요청 경로가 보호 대상이면 헤더 값과 저장된 시크릿을 비교한다.
 */
public class InternalSecretFilter extends OncePerRequestFilter {

    private static final Logger log = LoggerFactory.getLogger(InternalSecretFilter.class);

    private final InternalSecurityProperties properties;
    private final AntPathMatcher matcher = new AntPathMatcher();

    public InternalSecretFilter(InternalSecurityProperties properties) {
        this.properties = properties;
    }

    @Override
    protected void doFilterInternal(
            HttpServletRequest request,
            HttpServletResponse response,
            FilterChain filterChain) throws ServletException, IOException {

        if (!properties.isEnabled() || !isProtectedPath(request)) {
            filterChain.doFilter(request, response);
            return;
        }

        String providedSecret = request.getHeader(properties.getHeaderName());
        if (!isSecretValid(providedSecret, properties.getSecret())) {
            log.warn("내부 시크릿 검증 실패. path={}, remote={}", request.getRequestURI(), request.getRemoteAddr());
            response.sendError(HttpServletResponse.SC_UNAUTHORIZED, "UNAUTHORIZED_INTERNAL_CALL");
            return;
        }

        filterChain.doFilter(request, response);
    }

    private boolean isProtectedPath(HttpServletRequest request) {
        String uri = request.getRequestURI();
        return properties.getProtectedPathPatterns().stream()
                .filter(Objects::nonNull)
                .anyMatch(pattern -> matcher.match(pattern, uri));
    }

    private boolean isSecretValid(String providedSecret, String expectedSecret) {
        if (providedSecret == null || expectedSecret == null) {
            return false;
        }
        // 타이밍 공격을 줄이기 위해 해시 값을 비교한다.
        byte[] providedHash = sha256(providedSecret);
        byte[] expectedHash = sha256(expectedSecret);
        int result = 0;
        for (int i = 0; i < providedHash.length; i++) {
            result |= providedHash[i] ^ expectedHash[i];
        }
        return result == 0;
    }

    private byte[] sha256(String value) {
        try {
            MessageDigest digest = MessageDigest.getInstance("SHA-256");
            return digest.digest(value.getBytes(StandardCharsets.UTF_8));
        } catch (Exception ex) {
            throw new IllegalStateException("SHA-256 계산 실패", ex);
        }
    }
}

File: platform/common-lib/src/main/kotlin/com/researchex/common/security/InternalSecretFilter.kt
package com.researchex.common.security

import jakarta.servlet.FilterChain
import jakarta.servlet.ServletException
import jakarta.servlet.http.HttpServletRequest
import jakarta.servlet.http.HttpServletResponse
import org.slf4j.LoggerFactory
import org.springframework.util.AntPathMatcher
import org.springframework.web.filter.OncePerRequestFilter
import java.io.IOException
import java.nio.charset.StandardCharsets
import java.security.MessageDigest

/** 내부 호출용 시크릿 검증을 수행하는 필터다. 요청 경로가 보호 대상이면 헤더 값과 저장된 시크릿을 비교한다. */
class InternalSecretFilter(
    private val properties: InternalSecurityProperties
) : OncePerRequestFilter() {

    private val matcher = AntPathMatcher()

    @Throws(ServletException::class, IOException::class)
    override fun doFilterInternal(
        request: HttpServletRequest,
        response: HttpServletResponse,
        filterChain: FilterChain
    ) {
        if (!properties.enabled || !isProtectedPath(request)) {
            filterChain.doFilter(request, response)
            return
        }

        val providedSecret = request.getHeader(properties.headerName)
        if (!isSecretValid(providedSecret, properties.secret)) {
            log.warn("내부 시크릿 검증 실패. path={}, remote={}", request.requestURI, request.remoteAddr)
            response.sendError(HttpServletResponse.SC_UNAUTHORIZED, "UNAUTHORIZED_INTERNAL_CALL")
            return
        }

        filterChain.doFilter(request, response)
    }

    private fun isProtectedPath(request: HttpServletRequest): Boolean {
        val uri = request.requestURI
        return properties.protectedPathPatterns
            .filterNotNull()
            .any { pattern -> matcher.match(pattern, uri) }
    }

    private fun isSecretValid(providedSecret: String?, expectedSecret: String?): Boolean {
        if (providedSecret.isNullOrEmpty() || expectedSecret.isNullOrEmpty()) {
            return false
        }

        val providedHash = sha256(providedSecret)
        val expectedHash = sha256(expectedSecret)

        var result = 0
        for (index in providedHash.indices) {
            result = result or (providedHash[index].toInt() xor expectedHash[index].toInt())
        }
        return result == 0
    }

    private fun sha256(value: String): ByteArray {
        return try {
            MessageDigest.getInstance("SHA-256").digest(value.toByteArray(StandardCharsets.UTF_8))
        } catch (ex: Exception) {
            throw IllegalStateException("SHA-256 계산 실패", ex)
        }
    }

    companion object {
        private val log = LoggerFactory.getLogger(InternalSecretFilter::class.java)
    }
}
